src/buffer.cc

```diff
@@ -204,6 +205,82 @@ Result Buffer::CompareRMSE(Buffer* buffer, float tolerance) const {
   return {};
 }
 
+std::vector<double> Buffer::GetHistogramForChannel(uint32_t channel,
+                                                   uint32_t num_bins) const {
+  std::vector<double> bins(num_bins, 0.0);
+  auto* buf_ptr = GetValues<uint8_t>();
+  const auto& segments = format_->GetSegments();
+  const auto comp = segments[channel].GetComponent();
+  auto num_channels = format_->ValuesPerElement();
+  buf_ptr += channel * comp->SizeInBytes();
+  for (size_t i = 0; i < ElementCount(); ++i) {
+    const auto bin = *reinterpret_cast<const uint8_t*>(buf_ptr);
+    bins[bin]++;
+    // Assumes unpacked format where all channels have the same bit size
+    buf_ptr += comp->SizeInBytes() * num_channels;
+  }
+
+  return bins;
+}
+
+Result Buffer::CompareHistogramEMD(Buffer* buffer, float tolerance) const {
+  auto result = CheckCompability(buffer);
+  if (!result.IsSuccess())
+    return result;
+
+  const int num_bins = 256;
+  auto num_channels = format_->ValuesPerElement();
+
+  for (auto segment : format_->GetSegments()) {
+    if (!segment.GetComponent()->IsUint8() || num_channels != 4)
+      return Result(
+          "EMD comparison only supports 8bit unorm format with four channels.");
+  }
+
+  std::vector<std::vector<double>> histograms[2];
+  for (uint32_t c = 0; c < num_channels; ++c) {
+    histograms[0].push_back(GetHistogramForChannel(c, num_bins));
+    histograms[1].push_back(buffer->GetHistogramForChannel(c, num_bins));
+  }
+
+  // Normalize histograms
+  for (uint32_t i = 0; i < 2; ++i)
+    for (uint32_t c = 0; c < num_channels; ++c) {
+      double total = 0;
+      for (auto value : histograms[i][c])
+        total += value;
+      for (auto& value : histograms[i][c])
+        value /= total;
+    }
+
+  double max_emd = 0;
+
+  for (uint32_t c = 0; c < num_channels; ++c) {
+    std::vector<double> diffs(histograms[0][c].size());
+    for (size_t i = 0; i < histograms[0][c].size(); ++i)
+      diffs[i] = histograms[0][c][i] - histograms[1][c][i];
+    // Accumulate diffs
+    for (size_t i = 1; i < histograms[0][c].size(); ++i)
+      diffs[i] += diffs[i - 1];
```

Comment:
Can you add a comment before this method on what the algorithm is here? Why do we add the diffs together this way?

Reply:
We calculate the minimal cost of moving "earth" to transform the first histogram into the second, where each bin of the histogram can be thought of as a column of units of earth (i.e. dirt). The cost is the amount of earth moved times the distance carried (the distance is the number of adjacent bins over which the dirt is carried). The cumulative difference of the bins is one way to calculate this, as long as both histograms have the same amount of dirt. We have to sum the absolute values of the cumulative difference to get the final cost of how much (and how far) the dirt was moved.

Reply:
Nice comment, Paul.

"the distance is the number of adjacent bins over which the dirt is carried" - change "dirt" to "earth", as the only other mention of "dirt" is in an "i.e.", and I think it's best to have one term be the main term used.  (I would even drop the "i.e. dirt".)

"The cumulative difference of the bins is one way to calculate this" - phrase this to make it clear that this is the approach taken here: "We calculate this as the cumulative difference of ..."

"same amount of dirt" and "the dirt was moved": "dirt" -> "earth"